vms/avm/states/state.go

```diff
@@ -569,3 +599,224 @@ func (s *state) writeStatuses() error {
+	}
+
+	i := 1
+	for statusIter.Next() {
+		txIDBytes := statusIter.Key()
+		statusBytes := statusIter.Value()
+		if err := s.cleanupTx(lock, txIDBytes, statusBytes, txIter); err != nil {
+			return err
+		}
+
+		i++
+
+		if i%pruneCommitLimit == 0 {
+			// We must hold the lock during committing to make sure we don't
+			// attempt to commit to disk while a block is concurrently being
+			// accepted.
+			lock.Lock()
+			errs := wrappers.Errs{}
+			errs.Add(
+				s.Commit(),
+				statusIter.Error(),
+				txIter.Error(),
+			)
+			lock.Unlock()
+			if errs.Errored() {
+				return errs.Err
+			}
+
+			// We release the iterators here to allow the underlying database to
+			// clean up deleted state.
+			statusIter.Release()
+			txIter.Release()
+
+			now := time.Now()
+			if now.Sub(lastUpdate) > pruneUpdateFrequency {
+				lastUpdate = now
+
+				progress := timer.ProgressFromHash(txIDBytes)
+				eta := timer.EstimateETA(
+					startTime,
+					progress-startProgress,
+					stdmath.MaxUint64-startProgress,
+				)
+				log.Info("committing state pruning",
+					zap.Int("count", i),
+					zap.Duration("eta", eta),
+				)
+			}
+
+			// We take the minimum here because it's possible that the node is
+			// currently bootstrapping. This would mean that grabbing the lock
+			// could take an extremely long period of time; which we should not
+			// delay processing for.
+			pruneDuration := now.Sub(lastCommit)
+			sleepDuration := math.Min(
+				pruneCommitSleepMultiplier*pruneDuration,
+				10*time.Second,
+			)
+			time.Sleep(sleepDuration)
+
+			// Make sure not to include the sleep duration into the next prune
+			// duration.
+			lastCommit = time.Now()
+
+			// We shouldn't need to grab the lock here, but doing so ensures
+			// that we see a consistent view across both the statusDB and the
+			// txDB.
+			lock.Lock()
+			statusIter = s.statusDB.NewIteratorWithStart(txIDBytes)
+			txIter = s.txDB.NewIteratorWithStart(txIDBytes)
+			lock.Unlock()
+		}
+	}
+
+	lock.Lock()
+	defer lock.Unlock()
+
+	errs := wrappers.Errs{}
+	errs.Add(
+		s.Commit(),
+		statusIter.Error(),
+		txIter.Error(),
+	)
+
+	// Make sure we flush the original cache before re-enabling it to prevent
+	// surfacing any stale data.
+	oldTxCache.Flush()
+	s.statusesPruned = true
+	s.txCache = oldTxCache
+
+	log.Info("finished state pruning",
+		zap.Int("count", i),
+		zap.Duration("duration", time.Since(startTime)),
+	)
+
+	return errs.Err
+}
+
+// Assumes [lock] is unlocked.
+func (s *state) cleanupTx(lock sync.Locker, txIDBytes []byte, statusBytes []byte, txIter database.Iterator) error {
+	// After the linearization, we write txs to disk without statuses to mark
+	// them as accepted. This means that there may be more txs than statuses and
+	// we need to skip over them.
+	//
+	// Note: We do not need to remove UTXOs consumed after the linearization, as
+	// those UTXOs are guaranteed to have already been deleted.
+	if err := skipTo(txIter, txIDBytes); err != nil {
```

Comment:
If this returns an error, the loop will exit. Did we mean to gracefully handle the case where we can't skip ahead?

Reply:
There should never be a time that there is a status on disk and not a corresponding tx. So I think that should fail loudly.